[iOS] Fix thread safety issues when printing on a background thread

pxlcoder opened the pull request:
#### 7ab1a7bb64d82336e8fd6f64c49e00b85ba97302
<pre>
[iOS] Fix thread safety issues when printing on a background thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=242693">https://bugs.webkit.org/show_bug.cgi?id=242693</a>
rdar://93749291

Reviewed by Chris Dumez and Devin Rousso.

In iOS 16, UIKit added support for performing printing on a background thread
when using a UIPrintInteractionController. WebKit opts-in to background thread
printing when printing a web page, in order to avoid blocking the UI process.

Printing on a background thread means that the page count computation and drawing
are done off main thread. UIViewPrintFormatters must still be used on a single thread.

However, the existing logic has a couple of notable thread-safety issues:

1. Unsafe use of WebPageProxy in WKContentView (_WKWebViewPrintFormatter):

To fix this issue, all uses of `_page` are now performed on the main thread,
using `callOnMainRunLoop` and `ensureOnMainRunLoop`.

2. Unsynchronized, shared state representing the current print job on WKContentView:

WKContentView currently maintains a single printed document and a single
completion semaphore, even though multiple print jobs can be ongoing. There is
no synchronization between print jobs, which can result in multiple threads
reading and writing to the same variables representing the printing state.

Multiple background print jobs can be triggered using SPI, while a print job
on the main thread and a single background thread can be triggered using API.
In order to address the current thread-safety issues in this scenario, print
state is now maintained per-_WKWebViewPrintFormatter.

WKContentView maintains a set of waiting print formatters, in order to unblock
them should the web process get terminated.

* Source/WebKit/UIProcess/_WKWebViewPrintFormatter.mm:
(-[_WKWebViewPrintFormatter _printedDocument]):

Add a wrapper method to retrieve the printed document, ensuring the lock is
held.

(-[_WKWebViewPrintFormatter _setPrintedDocument:]):

Notify the waiting thread once the printed document has been retrieved.

(-[_WKWebViewPrintFormatter _waitForPrintedDocument]):
(-[_WKWebViewPrintFormatter _recalcPageCount]):
(-[_WKWebViewPrintFormatter drawInRect:forPageAtIndex:]):
* Source/WebKit/UIProcess/_WKWebViewPrintFormatterInternal.h:

Change `_wk_printedDocument` into `_wk_requestDocumentForPrintFormatter` as the
print formatter now contains state necessary for waiting.

* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _resetPrintingState]):

Wake up blocked print formatting threads as necessary.

(-[WKContentView _processDidExit]):
(-[WKContentView _frameIdentifierForPrintFormatter:]):

Factor out frame identifier retrieval into a common method to ensure it is
done on the main thread.

(-[WKContentView _wk_pageCountForPrintFormatter:]):

Move the frame identifier retrieval next to the page count computation and
perform the logic on the main thread.

`_page` can (incorrectly) be used off the main thread. To fix this, use
`ensureOnMainRunLoop` prior to calling WebPageProxy::drawToPDFiOS. Once the
PDF is retrieved, the background thread is woken up.

If printing on a background thread, add the print formatter to the pending set.

(-[WKContentView _waitForDrawToPDFCallbackForPrintFormatterIfNeeded:]):

Use the callback ID when printing on the main thread. On background threads,
consult the set of pending print formatters and wait as necessary.

(-[WKContentView _wk_requestDocumentForPrintFormatter:]):
(-[WKContentView _waitForDrawToPDFCallbackIfNeeded]): Deleted.
(-[WKContentView _wk_printedDocument]): Deleted.
* Source/WebKit/UIProcess/ios/WKPDFView.mm:
(-[WKPDFView _wk_requestDocumentForPrintFormatter:]):
(-[WKPDFView _wk_printedDocument]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewPrintFormatter.mm:
(TEST):

Added new tests to exercise the multiple background threads scenario, and the
main thread and single background thread scenario.

Canonical link: <a href="https://commits.webkit.org/252602@main">https://commits.webkit.org/252602@main</a>
</pre>

cdumez commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -148,9 +149,11 @@ @implementation WKContentView {
 
-    std::unique_ptr<BinarySemaphore> _pdfPrintCompletionSemaphore;
-    uint64_t _pdfPrintCallbackID;
+    Lock _pdfPrintCallbackMutex;
```
I think using Lock instead of Mutex in the variable name makes more sense, given the type.

cdumez commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -921,38 +926,72 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
+
+    ensureOnMainRunLoop([frameID, printInfo, pageCount, self, retainedSelf = retainPtr(self)] {
+        // Begin generating the PDF in expectation of a (eventual) request for the drawn data.
+        auto callbackID = _page->drawToPDFiOS(frameID, printInfo, pageCount, [retainedSelf = retainPtr(retainedSelf.get())](RefPtr<WebCore::SharedBuffer>&& pdfData) {
```
You should be able to capture just `retainedSelf`, no need for `= retainPtr(retainedSelf.get())`

cdumez commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -921,38 +926,72 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
+        if (!_pdfPrintCallbackID)
+            return true;
+
+        callbackID = *std::exchange(_pdfPrintCallbackID, std::nullopt);
```
The thing is that I don't think we really care about the callbackID in the case where we're on the background thread, since the `Condition` is sufficient then.

I think this code could probably be simplified like so:
```
if (isMainRunLoop()) {
    uint64_t callbackID;
    {
        Locker locker { _pdfPrintCallbackMutex };
        if (!_pdfPrintCallbackID)
            return true;
        callbackID = std::exchange(_pdfPrintCallbackID, 0);
    }
    return _page->process().connection()->waitForAsyncCallbackAndDispatchImmediately<Messages::WebPage::DrawToPDFiOS>(callbackID, Seconds::infinity());
}

Locker locker { _pdfPrintCallbackMutex };
if (_pdfPrintCompletionCondition) {
    _pdfPrintCompletionCondition->wait(_pdfPrintCallbackMutex);
return true;
```

And _pdfPrintCallbackID no longer needs to be a `std::optional<>`.

pxlcoder replied:
Thanks for the suggestion! I further simplified this now that `_pdfPrintCallbackID` is only accessed on the main thread.

cdumez commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -921,38 +926,72 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
+
+            Locker locker { retainedSelf->_pdfPrintCallbackMutex };
+
+            if (auto pdfPrintCompletionCondition = std::exchange(retainedSelf->_pdfPrintCompletionCondition, nullptr))
```
This code (nulling and notifying the condition + resetting the callbackID) is duplicated above so we may want to factor it out to its own function.

pxlcoder replied:
Done.

cdumez commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -921,38 +926,72 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
+            retainedSelf->_pdfPrintCallbackID = std::nullopt;
+        });
+
+        Locker locker { _pdfPrintCallbackMutex };
```
I don't think we really need locking for the callbackID if it is really only used from the main thread. We just need the lock for the Condition I believe.

pxlcoder replied:
Right, the callback ID is only used when printing from the main thread.

cdumez reviewed (approved):
LGTM

dcrousso commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -665,6 +668,15 @@ - (void)_webViewDestroyed
+
+    Locker locker { _pdfPrintCompletionLock };
+    if (auto pdfPrintCompletionCondition = std::exchange(_pdfPrintCompletionCondition, nullptr))
+        pdfPrintCompletionCondition->notifyAll();
```
Do we want to `notifyOne` here instead?  I would think that if we (somehow) have multiple prints queued up we'd only want to process one at a time.

pxlcoder replied:
I'm going to write a test to see whether we need to be concerned about this scenario (or if UIKit does the right thing on their own queue).

dcrousso commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -921,38 +930,45 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
-        retainedSelf->_pdfPrintCallbackID = 0;
+    ensureOnMainRunLoop([frameID, printInfo, pageCount, self, retainedSelf = retainPtr(self)] {
+        // Begin generating the PDF in expectation of a (eventual) request for the drawn data.
+        _pdfPrintCallbackID = _page->drawToPDFiOS(frameID, printInfo, pageCount, [retainedSelf](RefPtr<WebCore::SharedBuffer>&& pdfData) {
```
What about the `_page->mainFrame()` above?  Should that also only be called from the main thread?

pxlcoder replied:
Good catch – that will also need to be fixed.

dcrousso commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -921,38 +930,45 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
-            _pdfPrintCompletionSemaphore->wait();
-        else if (!_page->process().connection()->waitForAsyncCallbackAndDispatchImmediately<Messages::WebPage::DrawToPDFiOS>(callbackID, Seconds::infinity()))
-            return false;
+    if (isMainRunLoop()) {
```
What if the first print is done on a background thread, but then a second print is done on the main thread?  Shouldn't we still wait for the `_pdfPrintCompletionCondition`?

cdumez commented on Source/WebKit/UIProcess/_WKWebViewPrintFormatter.mm:
```diff
@@ -96,17 +128,19 @@ - (CGRect)rectForPageAtIndex:(NSInteger)pageIndex
-    if (!_printedDocument) {
-        _printedDocument = self._webView._printProvider._wk_printedDocument;
-        if (!_printedDocument)
+    CGPDFDocumentRef printedDocument = [self _printedDocument];
```
Would feel safer to retain the printedDocument

pxlcoder replied:
Done.

cdumez commented on Source/WebKit/UIProcess/_WKWebViewPrintFormatter.mm:
```diff
@@ -96,17 +128,19 @@ - (CGRect)rectForPageAtIndex:(NSInteger)pageIndex
-        if (!_printedDocument)
+    CGPDFDocumentRef printedDocument = [self _printedDocument];
+    if (!printedDocument) {
+        [self._webView._printProvider _wk_requestDocumentForPrintFormatter:self];
```
Feels like it'd be nice if _wk_requestDocumentForPrintFormatter would return the printed document instead of relying on a side effect from this function

pxlcoder replied:
I didn't make this change, because the implementation of `_wk_requestDocumentForPrintFormatter:` in the content view relies on the printed document getting set on the formatter as a side effect of async IPC. So, the "side effect" dependency would simply be moved from here into the content view.

Additionally, I would need to expose a `_printedDocument` property on `_WKWebViewPrintFormatter`. This is necessary as the printed document is now stored per-formatter. Finally, there would be duplicate code when returning from both the content view and PDF view cases (reaching back into the formatter).

I named the method `requestDocumentFor...` rather than `documentFor...` to indicate the lack of a return value.

dcrousso commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -148,9 +149,10 @@ @implementation WKContentView {
 
-    std::unique_ptr<BinarySemaphore> _pdfPrintCompletionSemaphore;
+    Lock _pendingBackgroundPrintFormattersLock;
+    RetainPtr<NSMutableSet> _pendingBackgroundPrintFormatters;
```
could/should we `WTF_GUARDED_BY_LOCK(_pendingBackgroundPrintFormattersLock)`?

cdumez replied:
I doubt this works in ObjC.

pxlcoder replied:
It doesn't, I tried it in an initial version of this patch.

dcrousso commented on Source/WebKit/UIProcess/_WKWebViewPrintFormatter.mm:
```diff
@@ -66,6 +71,33 @@ - (WKWebView *)_webView
+
+    Locker locker { _printLock };
+    _printedDocument = printedDocument;
+    _printCompletionCondition.notifyOne();
```
do we also need to do this in a `dealloc` or something in case `-[WKContentView _resetPrintingState]` is called?

pxlcoder replied:
UIKit will keep the print formatter alive until printing is completed, so it cannot be deallocated when waiting.

`-[WKContentView _resetPrintingState]` will unblock the thread in the event of web process termination.

dcrousso commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -861,6 +870,19 @@ @interface WKContentView (_WKWebViewPrintFormatter) <_WKWebViewPrintProvider>
+    if (auto mainFrame = _page->mainFrame())
+        return mainFrame->frameID();
+
+    return std::nullopt;
```
should we `ASSERT_NOT_REACHED`?

dcrousso commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -902,66 +916,100 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
+        [_pendingBackgroundPrintFormatters addObject:printFormatter];
+    }
+
+    ensureOnMainRunLoop([frameID, printInfo, pageCount, isPrintingOnBackgroundThread, printFormatter = retainPtr(printFormatter), self, retainedSelf = retainPtr(self)] {
```
NIT: is it necessary to also capture `self` if we already have `retainedSelf`?

pxlcoder replied:
It was necessary, because I was also accessing ivars directly. 

It was just in two locations, so I've removed the capture of `self` and added `retainedSelf->` where needed.

dcrousso commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -902,66 +916,100 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
+
+    ensureOnMainRunLoop([frameID, printInfo, pageCount, isPrintingOnBackgroundThread, printFormatter = retainPtr(printFormatter), self, retainedSelf = retainPtr(self)] {
+        // Begin generating the PDF in expectation of a (eventual) request for the drawn data.
+        auto callbackID = _page->drawToPDFiOS(frameID, printInfo, pageCount, [isPrintingOnBackgroundThread, printFormatter, retainedSelf](RefPtr<WebCore::SharedBuffer>&& pdfData) {
```
NIT: `printFormatter = WTFMove(printFormatter)`?

pxlcoder replied:
This results in the following error:

```
static_assert failed due to requirement '!is_const<const WTF::RetainPtr<_WKWebViewPrintFormatter>>::value' "T is const qualified."
```

dcrousso replied:
FWIW i think you can fix that by adding `mutable`, but is no biggie

dcrousso commented on Source/WebKit/UIProcess/_WKWebViewPrintFormatter.mm:
```diff
@@ -66,6 +71,33 @@ - (WKWebView *)_webView
 
+- (CGPDFDocumentRef)_printedDocument
+{
+    if (self.requiresMainThread)
```
It's a bit odd that `_WKWebViewPrintFormatter` exclusively checks for `requiresMainThread` while `WKWebView` only checks for `isMainRunLoop`.  Should we be checking `isMainRunLoop` here too in case `requiresMainThread` is `NO` but the `_WKWebViewPrintFormatter` is called on the main thread anyways?

pxlcoder replied:
No, we want to grab the lock, since PDF callback (writing the printed document) will always be called on the main thread, but the drawing can happen on a background thread (reading the printed document) .

We could optimize away the lock if the print formatter is used on the main thread only (the `UIPrintPageRenderer` case, which we can tell by where `_recalcPageCount` is called). However, since this is not a performance critical path, and the optimization would make the logic slightly more complex, I have kept the check simple. Printing a loaded PDF will avoid locking though.

dcrousso commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -902,66 +916,100 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
+
+    ensureOnMainRunLoop([frameID, printInfo, pageCount, isPrintingOnBackgroundThread, printFormatter = retainPtr(printFormatter), self, retainedSelf = retainPtr(self)] {
+        // Begin generating the PDF in expectation of a (eventual) request for the drawn data.
+        auto callbackID = _page->drawToPDFiOS(frameID, printInfo, pageCount, [isPrintingOnBackgroundThread, printFormatter, retainedSelf](RefPtr<WebCore::SharedBuffer>&& pdfData) {
```
what prevents this from being called multiple times on a non-main thread?  i would think/hope that subsequent requests to print would be delayed until the first request completes?

pxlcoder replied:
A couple of things:

1. The UIKit API itself is still main thread only, only UIKit internals can actually perform page count / drawing on a background thread.
2. The SPI to perform printing on a background thread needs a unique print formatter.
3. In the event (1) or (2) are no longer true, we still wait at the top of this method (`_waitForDrawToPDFCallbackForPrintFormatterIfNeeded`) if printing is ongoing, which delays subsequent requests.

pxlcoder committed:
[iOS] Fix thread safety issues when printing on a background thread
https://bugs.webkit.org/show_bug.cgi?id=242693
rdar://93749291

Reviewed by Chris Dumez and Devin Rousso.

In iOS 16, UIKit added support for performing printing on a background thread
when using a UIPrintInteractionController. WebKit opts-in to background thread
printing when printing a web page, in order to avoid blocking the UI process.

Printing on a background thread means that the page count computation and drawing
are done off main thread. UIViewPrintFormatters must still be used on a single thread.

However, the existing logic has a couple of notable thread-safety issues:

1. Unsafe use of WebPageProxy in WKContentView (_WKWebViewPrintFormatter):

To fix this issue, all uses of `_page` are now performed on the main thread,
using `callOnMainRunLoop` and `ensureOnMainRunLoop`.

2. Unsynchronized, shared state representing the current print job on WKContentView:

WKContentView currently maintains a single printed document and a single
completion semaphore, even though multiple print jobs can be ongoing. There is
no synchronization between print jobs, which can result in multiple threads
reading and writing to the same variables representing the printing state.

Multiple background print jobs can be triggered using SPI, while a print job
on the main thread and a single background thread can be triggered using API.
In order to address the current thread-safety issues in this scenario, print
state is now maintained per-_WKWebViewPrintFormatter.

WKContentView maintains a set of waiting print formatters, in order to unblock
them should the web process get terminated.

* Source/WebKit/UIProcess/_WKWebViewPrintFormatter.mm:
(-[_WKWebViewPrintFormatter _printedDocument]):

Add a wrapper method to retrieve the printed document, ensuring the lock is
held.

(-[_WKWebViewPrintFormatter _setPrintedDocument:]):

Notify the waiting thread once the printed document has been retrieved.

(-[_WKWebViewPrintFormatter _waitForPrintedDocument]):
(-[_WKWebViewPrintFormatter _recalcPageCount]):
(-[_WKWebViewPrintFormatter drawInRect:forPageAtIndex:]):
* Source/WebKit/UIProcess/_WKWebViewPrintFormatterInternal.h:

Change `_wk_printedDocument` into `_wk_requestDocumentForPrintFormatter` as the
print formatter now contains state necessary for waiting.

* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _resetPrintingState]):

Wake up blocked print formatting threads as necessary.

(-[WKContentView _processDidExit]):
(-[WKContentView _frameIdentifierForPrintFormatter:]):

Factor out frame identifier retrieval into a common method to ensure it is
done on the main thread.

(-[WKContentView _wk_pageCountForPrintFormatter:]):

Move the frame identifier retrieval next to the page count computation and
perform the logic on the main thread.

`_page` can (incorrectly) be used off the main thread. To fix this, use
`ensureOnMainRunLoop` prior to calling WebPageProxy::drawToPDFiOS. Once the
PDF is retrieved, the background thread is woken up.

If printing on a background thread, add the print formatter to the pending set.

(-[WKContentView _waitForDrawToPDFCallbackForPrintFormatterIfNeeded:]):

Use the callback ID when printing on the main thread. On background threads,
consult the set of pending print formatters and wait as necessary.

(-[WKContentView _wk_requestDocumentForPrintFormatter:]):
(-[WKContentView _waitForDrawToPDFCallbackIfNeeded]): Deleted.
(-[WKContentView _wk_printedDocument]): Deleted.
* Source/WebKit/UIProcess/ios/WKPDFView.mm:
(-[WKPDFView _wk_requestDocumentForPrintFormatter:]):
(-[WKPDFView _wk_printedDocument]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewPrintFormatter.mm:
(TEST):

Added new tests to exercise the multiple background threads scenario, and the
main thread and single background thread scenario.

Canonical link: https://commits.webkit.org/252602@main
